server/parsec/components/shamir.py

```diff
@@ -0,0 +1,195 @@
+                        )
+                    )
+                case RequireGreaterTimestamp() as error:
+                    return (
+                        authenticated_cmds.latest.shamir_recovery_setup.RepRequireGreaterTimestamp(
+                            strictly_greater_than=error.strictly_greater_than
+                        )
+                    )
+
+    # async def test_dump_current_shamir(
+    #     self, organization_id: OrganizationID
+    # ) -> dict[UserID, ShamirDump]:
+    #     raise NotImplementedError
+
+    async def remove_recovery_setup(
+        self,
+        organization_id: OrganizationID,
+        author: UserID,
+    ) -> (
+        None
+        | ShamirAddRecoverySetupStoreBadOutcome
+        | ShamirAddRecoverySetupValidateBadOutcome
+        | TimestampOutOfBallpark
+        | RequireGreaterTimestamp
+    ):
+        raise NotImplementedError
+
+    async def add_recovery_setup(
+        self,
+        organization_id: OrganizationID,
+        author: UserID,
+        device: DeviceID,
+        author_verify_key: VerifyKey,
+        setup: authenticated_cmds.latest.shamir_recovery_setup.ShamirRecoverySetup,
+    ) -> (
+        None
+        | ShamirAddRecoverySetupStoreBadOutcome
+        | ShamirAddRecoverySetupValidateBadOutcome
+        | TimestampOutOfBallpark
+        | RequireGreaterTimestamp
+    ):
+        raise NotImplementedError
+
+
+class ShamirAddRecoverySetupValidateBadOutcome(BadOutcomeEnum):
+    BRIEF_INVALID_DATA = auto()
+    SHARE_INVALID_DATA = auto()
+    SHARE_RECIPIENT_NOT_IN_BRIEF = auto()
+    DUPLICATE_SHARE_FOR_RECIPIENT = auto()
+    AUTHOR_INCLUDED_AS_RECIPIENT = auto()
+    MISSING_SHARE_FOR_RECIPIENT = auto()
+    THRESHOLD_GREATER_THAN_TOTAL_SHARES = auto()
+
+
+class ShamirAddRecoverySetupStoreBadOutcome(BadOutcomeEnum):
+    ORGANIZATION_NOT_FOUND = auto()
+    ORGANIZATION_EXPIRED = auto()
+    AUTHOR_NOT_FOUND = auto()
+    AUTHOR_REVOKED = auto()
+    INVALID_RECIPIENT = auto()
+    ALREADY_SET = auto()
+
+
+# Check internal consistency of certificate
+def shamir_add_recovery_setup_validate(
+    setup: authenticated_cmds.latest.shamir_recovery_setup.ShamirRecoverySetup,
+    author: DeviceID,
+    user_id: UserID,
+    author_verify_key: VerifyKey,
+) -> (
+    tuple[ShamirRecoveryBriefCertificate, dict[UserID, bytes]]
+    | ShamirAddRecoverySetupValidateBadOutcome
+    | TimestampOutOfBallpark
+):
+    share_certificates: dict[UserID, bytes] = {}
+    try:
+        brief_certificate = ShamirRecoveryBriefCertificate.verify_and_load(
+            setup.brief, author_verify_key, expected_author=author
+        )
+    except ValueError:
+        return ShamirAddRecoverySetupValidateBadOutcome.BRIEF_INVALID_DATA
+
+    match timestamps_in_the_ballpark(brief_certificate.timestamp, DateTime.now()):
+        case TimestampOutOfBallpark() as error:
+            return error
+        case _:
+            pass
```

Comment:
The timestamp for share certificates should also be checked.

We might also want to make sure that all those certificates share the same timestamp. That would probably help with the topic stuff.

Reply:
Nice catch!

Reply:
So if we check that each share timestamp is the same as the brief timestamp, it should be enough ?

Reply:
Yes, although you might have to yet another error status for that.